txnsync/incoming.go

```diff
@@ -0,0 +1,307 @@
+			// ( we have to disconnect, since otherwise, we would have no way to synchronize the sequence number)
+			s.log.Infof("unable to enqueue incoming message from a peer without txsync allocated data; incoming messages queue is full. disconnecting from peer.")
+			return errTransactionSyncIncomingMessageQueueFull
+		}
+		return nil
+	}
+	// place the incoming message on the *peer* heap, allowing us to dequeue it in the correct sequence order.
+	err = peer.incomingMessages.enqueue(incomingMessage)
+	if err != nil {
+		// if the incoming message queue for this peer is full, disconnect from this peer.
+		s.log.Infof("unable to enqueue incoming message into peer incoming message backlog. disconnecting from peer.")
+		return err
+	}
+
+	// (maybe) place the peer message on the main queue. This would get skipped if the peer is already on the queue.
+	enqueued := s.incomingMessagesQ.enqueue(incomingMessage)
+	if !enqueued {
+		// if we can't enqueue that, return an error, which would disconnect the peer.
+		s.log.Infof("unable to enqueue incoming message from a peer with txsync allocated data; incoming messages queue is full. disconnecting from peer.")
+		return errTransactionSyncIncomingMessageQueueFull
+	}
+	return nil
+}
+
+func (s *syncState) evaluateIncomingMessage(message incomingMessage) {
+	peer := message.peer
+	if peer == nil {
+		// check if a peer was created already for this network peer object.
+		peerInfo := s.node.GetPeer(message.networkPeer)
+		if peerInfo.NetworkPeer == nil {
+			// the message.networkPeer isn't a valid unicast peer, so we can exit right here.
+			return
+		}
+		if peerInfo.TxnSyncPeer == nil {
+			// we couldn't really do much about this message previously, since we didn't have the peer.
+			peer = makePeer(message.networkPeer, peerInfo.IsOutgoing, s.isRelay, &s.config)
+			// let the network peer object know about our peer
+			s.node.UpdatePeers([]*Peer{peer}, []interface{}{message.networkPeer}, 0)
+		} else {
+			peer = peerInfo.TxnSyncPeer
+		}
+		message.peer = peer
+		err := peer.incomingMessages.enqueue(message)
+		if err != nil {
+			// this is not really likely, since we won't saturate the peer heap right after creating it..
+			return
+		}
+	}
+	// clear the peer that is associated with this incoming message from the message queue, allowing future
+	// messages from the peer to be placed on the message queue.
+	s.incomingMessagesQ.clear(message)
+	messageProcessed := false
+	transactionPoolSize := 0
+	totalAccumulatedTransactionsCount := 0 // the number of transactions that were added during the execution of this method
+	transactionHandlerBacklogFull := false
+incomingMessageLoop:
+	for {
+		incomingMsg, seq, err := peer.incomingMessages.popSequence(peer.nextReceivedMessageSeq)
+		switch err {
+		case errHeapEmpty:
+			// this is very likely, once we run out of consecutive messages.
+			break incomingMessageLoop
+		case errSequenceNumberMismatch:
```

Comment:
it's a TCP connection, so that's not the case. The out of order is only due to the network library sending the messages to the transaction sync handler via a set of go-routines, which would re-order the messages.